# Ripassino
- Initializing a Tensor
    - torch.tensor()
    - array --> torch.from_numpy()
    - torch.ones_like()
    - torch.rand_like()
- Attributes of a tensor
    - shape
    - dtype
    - device
    - .to("mps")
- Operations on tensors:
    - slicing
    - indexing
    - .T
    - @
    - .matmul()
    - *

In [1]:
import numpy as np
import torch

In [3]:
matrix = np.array([
    [1,2,3],
    [4,5,6]
    ])

matrix

array([[1, 2, 3],
       [4, 5, 6]])

In [12]:
new_tensor = torch.from_numpy(matrix)
print(new_tensor.shape)

tensor = torch.tensor(data=[[1,3,5,7], [2,4,6,8]], dtype=torch.float64, requires_grad=True)
tensor

torch.Size([2, 3])


tensor([[1., 3., 5., 7.],
        [2., 4., 6., 8.]], dtype=torch.float64, requires_grad=True)

In [8]:
all_1_tensor = torch.ones_like(tensor, dtype=torch.int)

random_tensor = torch.rand_like(tensor)
random_tensor, all_1_tensor



(tensor([[0.3499, 0.2791, 0.0620, 0.9621],
         [0.3695, 0.8756, 0.6093, 0.6127]], dtype=torch.float64),
 tensor([[1, 1, 1, 1],
         [1, 1, 1, 1]], dtype=torch.int32))

In [13]:
print(tensor)
print(tensor.shape)
print(tensor.dtype)
print(tensor.device)

print()

tensor = tensor.to(dtype=torch.float32)
# print(tensor.dtype)
tensor = tensor.to('mps')   # from cpu --> gpu
tensor.device


tensor([[1., 3., 5., 7.],
        [2., 4., 6., 8.]], dtype=torch.float64, requires_grad=True)
torch.Size([2, 4])
torch.float64
cpu

torch.float32


device(type='mps', index=0)

---

# .backward()
- .backward() computes the **gradient**
- gradient = a function tells you how the function changes when you slightly change its inputs.
    - For a function of one variable → the gradient is just the derivative.
    - For a function of many variables → the gradient is the vector of partial derivatives.
    - Mathematically:
$ \nabla f(x_1, x_2, ..., x_n) = \left( \frac{\partial f}{\partial x_1}, \frac{\partial f}{\partial x_2}, \dots, \frac{\partial f}{\partial x_n} \right)$  

#### So the gradient is a vector that tells you the direction and rate of fastest increase of the function.  
Example:
- Suppose: $f(x,y) = x^2 + y^2$  
- The gradient is: $\nabla f = \left( \frac{\partial f}{\partial x}, \frac{\partial f}{\partial y} \right) = (2x, 2y)$
- At point (1,2): $\nabla f(1,2) = (2,4)$

Meaning:
- increasing x increases the function with slope 2
- increasing y increases the function with slope 4  
> So the function grows faster in the y direction.  

#### In poor words:
Imagine the function as a surface.  
The gradient vector at a point:
- points uphill
- points in the direction where the function increases the fastest
- its magnitude tells how steep the slope is  
> gradient = steepest uphill direction  

#### why do we care about the gradient?  
- Because we want to optimize functions.
- In machine learning we **minimize** a loss function: $L(\theta)$
    - where $\theta$ are the model parameters.
    - we said the gradient points into the direction the function increases --> we want to minimize --> go into the opposite direction of the gradient!
    - gradient tells us how to change the parameters to reduce the loss.

## Loss function  
- In machine learning we define a loss function: $L(\theta)$
- It measures how wrong the model is.


## Gradient descent  
- To minimize a function we move opposite the gradient.
- Why? Because the gradient points uphill --> So we go downhill.

# Update rule: $\theta_{new} = \theta - \eta \nabla L(\theta)$
where
- $\theta$ = parameters
- $\eta$ = learning rate
- $\nabla L(\theta)$ = gradient of the loss


> Example
- Loss function: $L(w) = (w-3)^2$  
- Derivative: $\frac{dL}{dw} = 2(w-3)$ --> So it is the slope of the loss curve.  
- If w = 10 then $\frac{dL}{dw} = 14$

Gradient descent step: $w_{new} = w - \eta \cdot 14$ --> meaning: the new weights are computed as the previous ones but moved towards the opposite direction of the gradient ... but how much are they're moved towards the gradient? It depends on the learning coefficient $\eta$ which decides if to move a lot towards the opposite direction of the gradient or just a little bit --> if to take a big step or a small step.  

But ... what the hell does gradient = 14 mean? IF WE MOVE TOWARDS RIGHT (positive values) / If we increase w slightly, the loss increases with slope 14: $\Delta L \approx 14 \cdot \Delta w$  
If we increase w by 0. --> $\Delta L \approx 14 \times 0.1 = 1.4$ --> So the loss increases.  

- increasing w → loss increases
- decreasing w → loss decreases
So to minimize the loss we must move left.

SO in 2D:
- Gradient > 0 --> if we move right (positive values) --> Loss increases --> we move left
- Gradient < 0 --> if we move left (negative values) --> Loss increases --> we move right

While in multiple dimensions n-D:
- the gradient really is a direction vector.
    - example in 2D:
$L(x,y) = (x-3)^2 + (y+1)^2$

The gradient is the vector of partial derivatives.  
$\nabla L = \left(\frac{\partial L}{\partial x},\frac{\partial L}{\partial y}\right)$  

Compute them:  
- derivative wrt x
$\frac{\partial L}{\partial x} = 2(x-3)$

- derivative wrt y
$\frac{\partial L}{\partial y} = 2(y+1)$

So  
$\nabla L = (2(x-3),\;2(y+1))$

- Suppose the parameters are (x,y) = (5,2)
    - First compute the loss: $L(5,2) = (5-3)^2 + (2+1)^2 = 4 + 9 = 13$
    - x component: 2(5-3)=4
    - y component: 2(2+1)=6
So $\nabla L = (4,6)$  

So
- increasing x increases the loss with slope 4
- increasing y increases the loss with slope 6
So moving up-right increases the loss.  
The gradient therefore points toward the steepest uphill direction.  

> Gradient descent  
- To minimize the loss we move opposite the gradient: $-\nabla L = (-4,-6)$
- Suppose the learning rate is $\eta = 0.1$
- Update rule: $(x_{new},y_{new}) = (x,y) - \eta \nabla L$
- Compute:
    - new x: $x_{new} = 5 - 0.1 \cdot 4 = 4.6$
    - new y: $y_{new} = 2 - 0.1 \cdot 6 = 1.4$
- So the parameters move: $(5,2) \rightarrow (4.6,1.4)$

In [19]:
# we cannot compute the gradient of a matrix, we can only compute the gradient of a scalar (number)
print(tensor)
tot_sum_scalar = tensor.sum()
scalar_function = (tensor**2).mean()

tot_sum_scalar, scalar_function

tensor([[1., 3., 5., 7.],
        [2., 4., 6., 8.]], device='mps:0', grad_fn=<ToCopyBackward0>)


(tensor(36., device='mps:0', grad_fn=<SumBackward0>),
 tensor(25.5000, device='mps:0', grad_fn=<MeanBackward0>))

- what gradient do we expect?
    - The sum function is something like: $L(T) = \sum_{i,j} T_{ij}$ --> 1+3+5+7+2+4+6+8 = 36  
    - Each of these summed values is a component so to compute the gradient we need to compute the partial derivate with repect each of these component: \frac{\partial L}{\partial T_{ij}}  
    - But the final scalar output can be written as the function **$L = T_{11} + T_{12} + T_{13} + ...$**  
    - So the partial derivatives are $\frac{\partial L}{\partial T_{ij}} = 1$ for every element  
THEREFORE the expected gradient is:
$\nabla L = \begin{bmatrix} 1 & 1 & 1 & 1 \\ 1 & 1 & 1 & 1 \end{bmatrix}$


- The second function is more complex:
    - First square every element: $T_{ij}^2$
    - Then take the mean
    - So the final scalar result can be written as: $L(T) = \frac{1}{8} \sum_{i,j} T_{ij}^2$

- Each variable is squared and in general the derivative of $x^2$ = 2x so each of these component can be replaced with $2T_{ij}$
- But the whole function is divided by 8, so:
$\frac{\partial L}{\partial T_{ij}} = \frac{2T_{ij}}{8} = \frac{T_{ij}}{4}$

THEREFORE the expected gradient is obtained dividing each tensor element by 4:
$\begin{bmatrix} 0.25 & 0.75 & 1.25 & 1.75 \\ 0.5 & 1.0 & 1.5 & 2.0 \end{bmatrix}$

---

# NOTE:
> .backward() computes gradients with respect to the variables that have requires_grad=True, NOT with respect to the final scalar itself.  
> Since we previously set in tensor: requires_grad=True --> when backpropagating / derivating from scalar_function we derivate with respect to tensor --> we compute the grad for tensor
$\frac{\partial scalar\_function}{\partial tensor}$

> #### Case 1  
Suppose the loss depends on two tensors:
- tensor1 → trainable parameter
- tensor2 → constant data

In [ ]:
import torch
tensor1 = torch.tensor(3.0, requires_grad=True)
tensor2 = torch.tensor(4.0, requires_grad=False)

scalar_function = tensor1**2 + tensor2**2

scalar_function.backward()

print(tensor1.grad)
print(tensor2.grad)


The loss is: $L = x^2 + y^2$  
where  
- x = tensor1
- y = tensor2

Derivative wrt x: $\frac{\partial L}{\partial x} = 2x$  
Derivative wrt y: $\frac{\partial L}{\partial y} = 2y$  

At $x=3,\quad y=4$  
$\frac{\partial L}{\partial x}=6$  
$\frac{\partial L}{\partial y}=8$  

**But** PyTorch only stores gradients for tensors requiring grad, so the result is:
- tensor(6.)
- None
    - because tensor2 was marked as constant.




> #### Case 2  
Now both are trainable.

In [ ]:
import torch

tensor1 = torch.tensor(3.0, requires_grad=True)
tensor2 = torch.tensor(4.0, requires_grad=True)

scalar_function = tensor1**2 + tensor2**2

scalar_function.backward()

print(tensor1.grad)
print(tensor2.grad)

Output:
- tensor(6.)
- tensor(8.)


The parameters are $\theta = (x,y)$  
Loss: $L(x,y) = x^2 + y^2$  

Gradient: $\nabla L = \left( \frac{\partial L}{\partial x}, \frac{\partial L}{\partial y} \right)$  
So $\nabla L = (6,8)$  
This is a direction vector in parameter space.  
Gradient descent would move in $-\nabla L = (-6,-8)$

In [24]:
tensor = torch.tensor(data=[[1,3,5,7], [2,4,6,8]], dtype=torch.float64, requires_grad=True)

tot_sum_scalar = tensor.sum()
scalar_function = (tensor**2).mean()

tot_sum_scalar.backward()

# NOT tot_sum_scalar.grad()
print(tensor.grad)

scalar_function.backward()
print(tensor.grad)


tensor([[1., 1., 1., 1.],
        [1., 1., 1., 1.]], dtype=torch.float64)
tensor([[1.2500, 1.7500, 2.2500, 2.7500],
        [1.5000, 2.0000, 2.5000, 3.0000]], dtype=torch.float64)


# OCCHIO! GRADIENTS GET AUTOMATICALLY SUMMED
- Pytorch computes the first gradient:
tensor.grad =  
[[1,1,1,1],  
 [1,1,1,1]]  

- It computes the second gradient:  
[[0.25,0.75,1.25,1.75],  
 [0.50,1.00,1.50,2.00]]  

> **But PyTorch ADDS it to the previous gradient.**  
> I need to reset the gradient: tensor.grad = None OR tensor.grad.zero_()

In [25]:
tensor = torch.tensor(data=[[1,3,5,7], [2,4,6,8]], dtype=torch.float64, requires_grad=True)

tot_sum_scalar = tensor.sum()
scalar_function = (tensor**2).mean()

tot_sum_scalar.backward()

# NOT tot_sum_scalar.grad()
print(tensor.grad)

# RESET GRADIENT, otherwise they silently get summed
tensor.grad.zero_()

scalar_function.backward()
print(tensor.grad)

tensor([[1., 1., 1., 1.],
        [1., 1., 1., 1.]], dtype=torch.float64)
tensor([[0.2500, 0.7500, 1.2500, 1.7500],
        [0.5000, 1.0000, 1.5000, 2.0000]], dtype=torch.float64)


---

---

#### Ok now we can continue with the lab

In [29]:
random_tensor = torch.rand_like(tensor)
print()

# slicing $ indexing [rows, columns]
print(random_tensor)
print(random_tensor[-1])    # last row and all columns
print(random_tensor[:, -1]) # all rows and last column
print(random_tensor[:,:-1]) # all rows and all columns except the last one
print()


# slicing 3d tensor
tensor_prova = tensor.reshape(shape=(2,2,2))
tensor_3d = torch.reshape(input=tensor, shape=(2,2,2))
print(tensor_3d)
print(tensor_3d.shape)

print(tensor_3d[1, 1, 0])
# take the second 2D matrix:
# [[2., 4.],
# [6., 8.]]
# of this 2D matrix take the second row:
# [6., 8.]
# and the first column:
# 6



tensor([[0.9117, 0.7035, 0.7915, 0.7637],
        [0.4769, 0.0877, 0.4497, 0.0079]], dtype=torch.float64)
tensor([0.4769, 0.0877, 0.4497, 0.0079], dtype=torch.float64)
tensor([0.7637, 0.0079], dtype=torch.float64)
tensor([[0.9117, 0.7035, 0.7915],
        [0.4769, 0.0877, 0.4497]], dtype=torch.float64)

tensor([[[1., 3.],
         [5., 7.]],

        [[2., 4.],
         [6., 8.]]], dtype=torch.float64, grad_fn=<ViewBackward0>)
torch.Size([2, 2, 2])
tensor(6., dtype=torch.float64, grad_fn=<SelectBackward0>)


---

# Datasets & DataLoaders

## Dataset
- The Dataset’s job: “If someone asks me for item in position 'i', I return the data in the row 'i' and the respective label in row 'i'.”  
Example:  

| index | image (data)    | label |
|------|----------|-------|
| 0    | cat.jpg  | 0     |
| 1    | dog.jpg  | 1     |
| 2    | bird.jpg | 2     |

The dataset simply return the row we ask:
- dataset[0] → (cat.jpg, 0)
- dataset[1] → (dog.jpg, 1)
- dataset[2] → (bird.jpg, 2)

So the Dataset answers: give me the i-th sample, return data AND label.  

Suppose the dataset is like:  

| X | Y |
|---|---|
| 1 | 0 |
| 2 | 1 |
| 3 | 0 |
| 4 | 1 |

If we had to manually build one it would be something like this:

In [ ]:
from torch.utils.data import Dataset

class MyDataset(Dataset):

    def __init__(self):
        self.data = [1,2,3,4]
        self.labels = [0,1,0,1]

    def __len__(self):
        return len(self.data)

    def __getitem__(self, idx):
        return self.data[idx], self.labels[idx]

In [ ]:
dataset = MyDataset()
dataset[0] # --> returns (1,0) --> take the first row and return the data (X) and its label (Y)

## Dataloader
The DataLoader’s job is different.  
It says: “Give me several samples from the dataset and pack them into batches.”  

**Batches = sottogruppi**

If the Dataset has: [1,2,3,4]  
And if batch size = 2  
Then DataLoader produces:  
[1,2]  
[3,4]  

> It 

> Why batches exist?  
- Neural networks do not train one sample at a time.  
- Instead they process small groups of samples.  
    - If Dataset size: 60,000 images --> I cannot train the NN on the whole dataset once --> I divide the 60,000 in sub-groups and train on each sub-group
    - Batch size: 64
    - Training loop receives:
        - 64 images
        - 64 images
        - 64 images
        - ...
    - instead of loading all 60k images into memory.

> Analogy: do you think it's better studying reading all 500 pages and then repeat or read 20 pages at the time and then repeat? Same for a NN, it learns better.

> Also a Dataloader allows to:
- Shuffle data
- Parallel loading

# WHEN WE ITERATE ON THE DATA LOADER IT ITERATES ON 2 ITEMS: DATA, LABELS

In [ ]:
from torch.utils.data import DataLoader
dataloader = DataLoader(dataset, batch_size=32, shuffle=True, num_workers=4)

for images, labels in dataloader:
    pass

# Training loop + evaluation loop example

In [ ]:
import torch
import torch.nn as nn
from torch.utils.data import Dataset, DataLoader


# =========================================
# 1) DATASET
# =========================================
# Job of the Dataset:
# "Given an index i, return ONE sample -> (data, label)"

class MyDataset(Dataset):
    def __init__(self):
        # Fake dataset: 100 grayscale images of shape [1, 28, 28]
        self.images = torch.randn(100, 1, 28, 28)
        self.labels = torch.randint(0, 10, (100,))   # classes 0,...,9

    def __len__(self):
        return len(self.images)

    def __getitem__(self, idx):
        image = self.images[idx]   # shape: [1, 28, 28]
        label = self.labels[idx]   # scalar
        return image, label


# =========================================
# 2) DATALOADER
# =========================================
# Job of the DataLoader:
# "Given the whole Dataset build sub-groups = batches"

train_dataset = MyDataset()
train_loader = DataLoader(train_dataset, batch_size=16, shuffle=True)

val_dataset = MyDataset()
val_loader = DataLoader(val_dataset, batch_size=16, shuffle=False)


# =========================================
# 3) MODEL
# =========================================
# The __init__ method specifies all the tools you need in the forward method, which is the workflow of the Neural Network, the journey of the input / how the input data flows through the layers.
# In the __init__ we initialise the layers + the activation functions (once)

# The forward method is the workflow:
# Take an input which has shape (N, 784) --> has dimensionality (number of columns) = 784
# --> multiply it by a weight matrix W (784, 512) --> the output becomes (N, 512), has dimensionality 512 --> reduce its dimensionality to 512
# what actually happens in each layer is: Y = X @ W.T + b

# Apply the activation function (non-linear function) ReLu --> maps every value to max(0, x) --> maps negative values to 0
# Then the reduced and ReLued input gets reduced from dimensionality 512 to 256
# ReLu again
# Then reduced again to 10 columns


class SimpleNN(nn.Module):
    def __init__(self):
        super(SimpleNN, self).__init__()

        self.layer_1 = nn.Linear(784, 512)
        self.relu = nn.ReLU()
        self.layer_2 = nn.Linear(512, 256)
        self.layer_3 = nn.Linear(256, 10)

    def forward(self, x):
        # x arrives as a batch of images:
        # [batch_size, 1, 28, 28]
        #
        # We flatten each image:
        # [batch_size, 1, 28, 28] -> [batch_size, 784]
        x = x.reshape(x.shape[0], 784)

        x = self.layer_1(x)
        x = self.relu(x)

        x = self.layer_2(x)
        x = self.relu(x)

        x = self.layer_3(x)   # output logits, shape [batch_size, 10]
        return x


# =========================================
# 4) DEVICE
# =========================================
# On Apple Silicon use "mps" if available, otherwise CPU

if torch.backends.mps.is_available():
    device = torch.device("mps")
else:
    device = torch.device("cpu")


# =========================================
# 5) LOSS + OPTIMIZER
# =========================================

model = SimpleNN().to(device)

# Since this is a 10-class classification problem:
# output shape: [batch_size, 10]
# labels shape: [batch_size]
loss_fn = nn.CrossEntropyLoss()

# Adam optimizer
optimizer = torch.optim.Adam(model.parameters(), lr=1e-3)


# =========================================
# 6) TRAINING LOOP
# =========================================
# 1 epoch = we see the whole dataset 1 time but divided in batches --> epoch = 3 --> we see the whole dataset 3 times but 'spezzettato' in many batches
# for each epoch 



num_epochs = 3

for epoch in range(num_epochs):
    # TRAIN MODE
    model.train()

    running_loss = 0.0

    for images, labels in train_loader:
        # images: [batch_size, 1, 28, 28]
        # labels: [batch_size]

        images = images.to(device)
        labels = labels.to(device)

        # !!!!!!! VERY IMPORTANT !!!!!!!
        # clear old gradients, otherwise PyTorch accumulates them
        optimizer.zero_grad()

        # forward pass
        outputs = model(images)   # [batch_size, 10]

        # compute loss
        loss = loss_fn(outputs, labels)

        # backward pass
        loss.backward()

        # update parameters
        optimizer.step()

        running_loss += loss.item()

    avg_train_loss = running_loss / len(train_loader)
    print(f"Epoch {epoch+1} - Train loss: {avg_train_loss:.4f}")


Before explaining the evaluation part we need to understand what are.  

# Logits
- After the forward pass:
outputs = model(images) --> shape: [batch_size, 10] because the last layer of the NN has 10 outputs  
SO for each image, the network produces 10 numbers. Why 10 numbers? Because we suppose that in our example there are 10 classes, so one output per class.  
The outputs can be something like this:  
outputs = [1.2, -0.5, 0.7, 2.3, -1.0, 0.1, -0.3, 0.8, -0.2, 0.4]  
> These numbers are called **logits**.

Logits are raw scores produced by the neural network before converting them to probabilities.  
They are not probabilities:  
- they can be negative
- they do not sum to 1

| class  | 0   | 1    | 2   | **3** | 4    | 5   | 6    | 7   | 8    | 9   |
|--------|-----|------|-----|-------|------|-----|------|-----|------|-----|
| logits | 1.2 | -0.5 | 0.7 | **2.3** | -1.0 | 0.1 | -0.3 | 0.8 | -0.2 | 0.4 |

They're not probabilities but almoust, so **the largest score indicates the class the network thinks is most likely**.  
In this case:  
- max logit = 2.3
- index = 3
SO: the predicted class is 3

# SOFTMAX FUNCTION
But there is an additional step because we are using a special loss function called: **CrossEntropyLoss**  
This function internally applies an activation function by itself --> **softmax function**:
  
$P(class_i) = \frac{e^{logit_i}}{\sum_j e^{logit_j}}$  

This activation function transforms the logits into probabilities  
So why the hell did we previously 'frettolosamente' pick the max logit = 2.3 and did not convert into a probability first using the softmax formula? Because Softmax preserves the order of logits --> the max logits remains the max value even after converting it into a probability :)  

So instead of computing softmax, we can directly do: predicted class = index of largest logit

In [ ]:
num_epochs = 3

for epoch in range(num_epochs):
    
    # =========================================
    # 7) EVALUATION LOOP
    # =========================================
    model.eval()

    correct = 0
    total = 0
    val_loss = 0.0

    # During evaluation we do NOT need gradients, we don't want the optimizer to make our predictions better
    with torch.no_grad():
        for images, labels in val_loader:
            images = images.to(device)
            labels = labels.to(device)

            outputs = model(images)          # [batch_size, 10]
            loss = loss_fn(outputs, labels)
            val_loss += loss.item()

            # This is the hard part
            # Predicted class = index of max logit
            preds = torch.argmax(outputs, dim=1)

            num_correct = (preds == labels).sum().item()
            
            batch_size = labels.shape[0]
            
            total += batch_size
            correct += num_correct
            
    avg_val_loss = val_loss / len(val_loader)
    accuracy = correct / total

    print(f"Epoch {epoch+1} - Val loss: {avg_val_loss:.4f} - Val acc: {accuracy:.4f}")

- Output have shape: [batch_size, 10]
	- each row = one image
	- each column = score for one class (0–9)
    - example:
outputs =  
[  
 [1.2, -0.5, 0.7, 2.3, -1.0, 0.1, -0.3, 0.8, -0.2, 0.4],  
 [0.2,  0.1, 3.1, 0.5, -0.7, 0.2, -0.4, 0.1,  0.0, 0.3],  
 [-1.0, 2.2, 0.3, 0.1, -0.5, 0.2, 0.0, -0.1, 0.4, 0.7]  
]  

- we pick the predicted class:
    - preds = torch.argmax(outputs, dim=1)
    - So for each row look for the biggest value and pick its position --> preds = [3, 2, 1]
    - meaning:
        - image 1 → predicted class 3
        - image 2 → predicted class 2
        - image 3 → predicted class 1


- preds == labels
    - preds  = [3, 2, 1]
    - labels = [3, 2, 4]
    - result: [True, True, False]

- (preds == labels).sum()
    - returns 2
    - But PyTorch returns a tensor, so we convert it to a normal number with .item()

- num_correct = (preds == labels).sum().item()
    - number of correct predictions in this batch.